pkg/controllers/hub/trafficmanagerprofile/controller.go

```diff
@@ -0,0 +1,271 @@
+
+	azureProfileName := generateAzureTrafficManagerProfileNameFunc(profile)
+	klog.V(2).InfoS("Deleting Azure Traffic Manager profile", "trafficManagerProfile", profileKObj, "azureProfileName", azureProfileName)
+	if _, err := r.ProfilesClient.Delete(ctx, r.ResourceGroupName, azureProfileName, nil); err != nil {
+		if !azureerrors.IsNotFound(err) {
+			klog.ErrorS(err, "Failed to delete Azure Traffic Manager profile", "trafficManagerProfile", profileKObj, "azureProfileName", azureProfileName)
+			return ctrl.Result{}, err
+		}
+	}
+
+	controllerutil.RemoveFinalizer(profile, objectmeta.TrafficManagerProfileFinalizer)
+	if err := r.Client.Update(ctx, profile); err != nil {
+		klog.ErrorS(err, "Failed to remove trafficManagerProfile finalizer", "trafficManagerProfile", profileKObj)
+		return ctrl.Result{}, err
+	}
+	klog.V(2).InfoS("Removed trafficManagerProfile finalizer", "trafficManagerProfile", profileKObj)
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) handleUpdate(ctx context.Context, profile *fleetnetv1alpha1.TrafficManagerProfile) (ctrl.Result, error) {
+	profileKObj := klog.KObj(profile)
+	azureProfileName := generateAzureTrafficManagerProfileNameFunc(profile)
+	var responseError *azcore.ResponseError
+	getRes, getErr := r.ProfilesClient.Get(ctx, r.ResourceGroupName, azureProfileName, nil)
+	if getErr != nil {
+		if !azureerrors.IsNotFound(getErr) {
+			klog.ErrorS(getErr, "Failed to get the profile", "trafficManagerProfile", profileKObj, "azureProfileName", azureProfileName)
+			return ctrl.Result{}, getErr
+		}
+		klog.V(2).InfoS("Azure Traffic Manager profile does not exist", "trafficManagerProfile", profileKObj, "azureProfileName", azureProfileName)
+	} else {
+		existingSpec := convertToTrafficManagerProfileSpec(&getRes.Profile)
+		if equality.Semantic.DeepEqual(existingSpec, profile.Spec) {
+			// skip creating or updating the profile
+			klog.V(2).InfoS("No profile update needed", "trafficManagerProfile", profileKObj, "azureProfileName", azureProfileName)
+			return r.updateProfileStatus(ctx, profile, getRes.Profile, nil)
+		}
+	}
+
+	res, updateErr := r.ProfilesClient.CreateOrUpdate(ctx, r.ResourceGroupName, azureProfileName, generateAzureTrafficManagerProfile(profile), nil)
+	if updateErr != nil {
+		if !errors.As(updateErr, &responseError) {
+			klog.ErrorS(updateErr, "Failed to send the createOrUpdate request", "trafficManagerProfile", profileKObj, "azureProfileName", azureProfileName)
+			return ctrl.Result{}, updateErr
+		}
+		klog.ErrorS(updateErr, "Failed to create or update a profile", "trafficManagerProfile", profileKObj,
+			"azureProfileName", azureProfileName,
+			"errorCode", responseError.ErrorCode, "statusCode", responseError.StatusCode)
+	}
+	return r.updateProfileStatus(ctx, profile, res.Profile, updateErr)
+}
+
+func convertToTrafficManagerProfileSpec(profile *armtrafficmanager.Profile) fleetnetv1alpha1.TrafficManagerProfileSpec {
+	if profile.Properties != nil && profile.Properties.MonitorConfig != nil {
```

Comment:
Do we want to return or log an error if the Properties or MonitorConfig or Protocol is nil? These should not happen because we have default values right?

Reply:
It depends on how Azure traffic manager handle it and we cannot make the assumption. For example, another client sets the properties to nil at the same time.